### DISCUSSION TITLE
Saving the WebGPU canvas to an image file

### DISCUSSION THREAD

**Discussion by subblue** [gpuweb/gpuweb#4493] [Q&A] [answered]
Looking through the issues there are various discussions (e.g. #2743) on `toBlob()`, `toDataURL()` etc for getting the rendered WebGPU canvas data, for instance to save a copy of the image to disc - a key requirement for my app. 
From WebGL we're familiar with the `preserveDrawingBuffer: true` approach.

The explainer has a section about [Current Texture](https://gpuweb.github.io/gpuweb/explainer/#canvas-output-current-texture), but it still isn't clear on how to make use of this in practise to snapshot the current canvas image, i.e:

* Do we have to inject something into our requestAnimationFrame loop?
* Should that exist within a `device.queue.onSubmittedWorkDone().then(...)` promise?
* How is `context.getCurrentTexture()` used to get the data in a form suitable for `canvas.toDataURL()`, `canvas.toBlob()` etc?

Any clarifications would be much appreciated!

**Comment by subblue** [comment, marked as answer]
In the process of asking this question I then realised the answer, as can happen!
For others that might have the same need:

Within the render loop there is something like this

```ts
device.queue.onSubmittedWorkDone()
  .then(() => {
    onCapture?.(canvas)
  })
```

where the `onCapture()` callback is set conditionally only when an image export is requested.
Within that the canvas image data is still available for saving before the next frame resets it, e.g:

```ts
function onCapture(canvas) {
  canvas.toBlob((blob) => { ... }, mimetype, quality)
  ...
}
```

**Reply by Kangz** [reply]
If you want to capture the current frame, you shouldn't need to wait for submitted work to be done, and instead IIRC the browser will do the wait for you. ToBlob will capture the last drawn frame so you might be getting an image that's a couple frames after the frame you think the code is capturing.
